File: hard/072_Edit_Distance/edit_distance.c
#include <stddef.h>
#include <stdlib.h>
#include <string.h>

#define MIN(a, b) ((a) < (b) ? (a) : (b))

static int
min_distance_internal (char *s, int len_s, char *t, int len_t)
{
    int *dp, result, i, j, prev, temp;

    dp = calloc(len_t+1, sizeof(int));
    if (dp == NULL) {
        return -1;
    }

    for (i = 0; i <= len_t; i++) {
        dp[i] = i;
    }

    for (i = 1; i <= len_s; i++) {
        prev = dp[0];
        dp[0] += 1;
        for (j = 1; j <= len_t; j++) {
            temp = dp[j];
            if (s[i-1] == t[j-1]) {
                dp[j] = prev;
            } else {
                dp[j] = MIN(dp[j-1]+1, dp[j]+1);
                dp[j] = MIN(prev+1, dp[j]);
            }
            prev = temp;
        }
    } 

    result = dp[len_t];
    free(dp);
    return result;
}

int
minDistance (char *s, char *t)
{
    int len_s, len_t, result;

    if (s == NULL || t == NULL) {
        return 0;
    }

    len_s = strlen(s);
    len_t = strlen(t);
    if (len_s >= len_t) {
        result = min_distance_internal(s, len_s, t, len_t);
    } else {
        result = min_distance_internal(t, len_t, s, len_s);
    }
    return result;
}

#include <stdio.h>

static void
test_min_distance (char *s, char *t)
{
    int result;

    result = minDistance(s, t);
    printf("%s, %s, min distance %d\n", s, t, result);
    return;
}

int main (void)
{
    test_min_distance("", "");
    test_min_distance("", "a");
    test_min_distance("xxxab", "abyyy");
    return 0;
}

File: hard/072_Edit_Distance/edit_distance.py
class Solution(object):
    def minDistance(self, s, t):
        """
        :type word1: str
        :type word2: str
        :rtype: int
        """
        if len(s) >= len(t):
            r = self.min_distance_internal(s, t)
        else:
            r = self.min_distance_internal(t, s)
        return r

    def min_distance_internal(self, s, t):
        dp = [i for i in xrange(len(t)+1)]
        for i in xrange(1, len(s)+1):
            prev = dp[0]
            dp[0] += 1
            for j in xrange(1, len(t)+1):
                temp = dp[j]
                if s[i-1] == t[j-1]:
                    dp[j] = prev
                else:
                    dp[j] = min(dp[j-1]+1, dp[j]+1)
                    dp[j] = min(dp[j], prev+1)
                prev = temp
        return dp[-1]

def test_edit_distance(a, b):
    r = solution.minDistance(a, b)
    print "{}, {}, edit distance {}".format(a, b, r)

solution = Solution()

a = ""
b = ""
test_edit_distance(a, b)

a = ""
b = "a"
test_edit_distance(a, b)

# Why could not apply LCS, see this example
a = "xxxab"
b = "abyyy"
test_edit_distance(a, b)
